src/DynamoCore/Core/CustomNodeManager.cs

```diff
@@ -613,26 +658,35 @@ internal bool TryGetInfoFromPath(string path, bool isTestMode, out CustomNodeInf
 
         private bool InitializeCustomNode(
             WorkspaceInfo workspaceInfo,
-            XmlDocument xmlDoc, out CustomNodeWorkspaceModel workspace)
+            out CustomNodeWorkspaceModel workspace,
+            Engine.LibraryServices libraryServices)
         {
             // Add custom node definition firstly so that a recursive
             // custom node won't recursively load itself.
             SetPreloadFunctionDefinition(Guid.Parse(workspaceInfo.ID));
- 
-            var nodeGraph = NodeGraph.LoadGraphFromXml(xmlDoc, nodeFactory);
-           
-            var newWorkspace = new CustomNodeWorkspaceModel(
+
+            XmlDocument xmlDoc;
+            string jsonDoc;
+            CustomNodeWorkspaceModel newWorkspace = null;
+            if (DynamoUtilities.PathHelper.isValidXML(workspaceInfo.FileName, out xmlDoc))
+            {
+                var nodeGraph = NodeGraph.LoadGraphFromXml(xmlDoc, nodeFactory);
+                newWorkspace = new CustomNodeWorkspaceModel(
                 nodeFactory,
                 nodeGraph.Nodes,
                 nodeGraph.Notes,
                 nodeGraph.Annotations,
-                nodeGraph.Presets,              
+                nodeGraph.Presets,
                 nodeGraph.ElementResolver,
                 workspaceInfo);
+            }
+            else if(DynamoUtilities.PathHelper.isValidJson(workspaceInfo.FileName, out jsonDoc))
```

Comment:
where these cases are for JSON, can you add a comment specifying that... or even break out the logic into a function like.... doingWhateverForJson... just a thought not required.

Reply:
hmm.. What do you mean?